compiler/src/iree/compiler/DispatchCreation/SetEncoding.cpp

```diff
@@ -227,17 +229,45 @@ class SetContractionOpEncoding final
     SmallVector<AffineMap> maps = linalgOp.getIndexingMapsArray();
 
     auto opType = IREE::Encoding::EncodingOpType::matmul;
-    auto setEncodingWrapper = [&](Value src, int64_t operandIndex) -> Value {
-      auto encoding =
-          EncodingAttr::get(linalgOp.getContext(), operandIndex, opType,
-                            elemTypes, maps, iterationSizes);
+    auto setEncodingWrapper = [&](Value src,
+                                  int64_t operandIndex) -> FailureOr<Value> {
+      MLIRContext *ctx = linalgOp.getContext();
+      Attribute encoding;
+      switch (encodingOption) {
+      case EncodingOptions::Default: {
+        encoding = EncodingAttr::get(ctx, operandIndex, opType, elemTypes, maps,
+                                     iterationSizes);
+        break;
+      }
+      case EncodingOptions::MatmulK: {
+        SmallVector<int32_t> kDims;
+        AffineMap indexingMap = maps[operandIndex];
+        auto cDims = linalg::inferContractionDims(linalgOp);
+        for (auto k : cDims->k) {
+          std::optional<unsigned> dimIdx =
+              indexingMap.getResultPosition(rewriter.getAffineDimExpr(k));
+          if (!dimIdx) {
+            continue;
+          }
+          kDims.push_back(dimIdx.value());
+        }
+        encoding = MatmulKAttr::get(ctx, rewriter.getDenseI32ArrayAttr(kDims));
+        break;
+      }
+      default: {
+        return failure();
```

Comment:
It is very wrong when we reach to the code and we don't need error handling in this case. We can just do an assert(false) and return `Value()`. I'd avoid `FailureOr<>` if there is no difference between NULL and a failure.